functions/src/plugins/size.ts

```diff
@@ -0,0 +1,268 @@
+
+    const failure = this.isFailure(config, largestIncrease.increase);
+
+    if(failure) {
+      const desc = `${largestIncrease.artifact.fullPath} increased by ${largestIncrease.increase} bytes`; // TODO pretty up bytes 
+      await this.setStatus(STATUS_STATE.Failure, desc, context, config);
+    } else {
+      if(largestIncrease.increase === 0) {
+        const desc = `no size change`;
+        await this.setStatus(STATUS_STATE.Success, desc, context, config);      
+      } else if (largestIncrease.increase < 0) {
+        const desc = `${largestIncrease.artifact.fullPath} decreased by ${largestIncrease.increase} bytes`; // TODO pretty up bytes 
+        await this.setStatus(STATUS_STATE.Success, desc, context, config); 
+      }
+    }
+  }
+
+  async findCurrentPr(sha: string, repositoryId: number) {
+    let pr;
+    const matches = (await this.pullRequests.where('head.sha', '==', sha)
+      .where('repository.id', '==', repositoryId)
+      .get());
+    matches.forEach(async doc => {
+      pr = doc.data();
+    });
+    return pr;
+  }
+
+  async storeArtifacts(context: Context) {
+    const {owner, repo} = context.repo();
+    const buildNumber = await this.getBuildNumberFromCircleCIUrl(context.payload.target_url);
+    const newArtifacts = await this.getCircleCIArtifacts(owner, repo, buildNumber);
+    await this.upsertNewArtifacts(context, newArtifacts);
+  }
+
+  async upsertNewArtifacts(context: Context, artifacts: BuildArtifact[]) {
+    // eg: aio/gzip7/inline
+    // eg: ivy/gzip7/inline
+    // projects within this repo 
+    const projects = new Set(artifacts.map(a => a.projectName));
+
+    for(const project of projects) {
+      for(const branch of context.payload.branches) {
+
+        const ref =  this.rtDb.ref(`/payload/${project}/${branch.name}/${context.payload.commit.sha}`);
+        const artifactsOutput = {
+          change: 'application',
+          message: context.payload.commit.commit.message,
+          timestamp: new Date().getTime(),
+        };
+  
+        // only use the artifacts from this project
+        artifacts.filter(a => a.projectName === project)
+          .forEach(a => {
+            // hold a ref to where we are in our tree walk
+            let lastNestedItemRef: object|number = artifactsOutput;
+            // first item is the project name which we've used already 
+            a.contextPath.forEach((path, i) => {
+              // last item so assign it the bytes size
+              if(i === a.contextPath.length - 1) {
+                lastNestedItemRef[path] = a.sizeBytes;
+                return;
+              }
+              if(!lastNestedItemRef[path]) {
+                lastNestedItemRef[path] = {};
+              }
+  
+              lastNestedItemRef = lastNestedItemRef[path];
+            });
+            lastNestedItemRef = a.sizeBytes;
+          });     
+        // if one already exists for this sha, override it
+        await ref.set(artifactsOutput);
```

Comment:
Since it's a async function and im using `for ... of ...` each iteration occurs in sequence and the Promise returned does not resolve  till all iterations are done